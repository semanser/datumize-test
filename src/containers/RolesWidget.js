import React, { Component } from 'react'
import { connect } from 'react-redux'
import { GridLoader } from 'halogenium'

import * as actions from '../redux/actions'

import RolesSelect from '../components/RolesSelect'
import RoleItem from '../components/RoleItem'
import UserItem from '../components/UserItem'

export class RolesWidget extends Component {
  componentDidMount() {
    this.props.fetchData()
  }

  render() {
    if (this.props.isFetching) {
      return (
        <div className="widget">
          <GridLoader color="#5E85E7" size="10px" margin="2px" className="loader"/>
        </div>
      )
    }

    const currentProject = this.props.projects.find(p => p.id === this.props.selectedProjectId)

    return (
      <div className="widget">
        {this.props.mode === 'default'
            ? <RolesSelect {...this.props}/>
            : <div className="header">
              <div className="title">{currentProject.name}</div>
              <div className="description">
                {this.props.mode === 'select-user'
                    ? 'Select the user below'
                    : `Now select the role for ${this.props.users.find(user => user.id === this.props.selectedUserId).name}` }
              </div>
            </div>
        }     

        {this.props.mode === 'default' &&
            this.props.filteredUsers.map(user =>
              <UserItem
                user={user}
                onUnassign={this.props.unassignUser.bind(this, user.id, this.props.selectedProjectId)}
                key={user.id}
              />
            )
        }

        {this.props.mode === 'select-user' &&
            this.props.filteredUsers.map(user =>
              <UserItem
                user={user}
                onSelect={this.props.selectUser.bind(this, user.id)}
                key={user.id}
              />
            )
        }

        {this.props.mode === 'select-role' &&
            this.props.roles.map(role =>
              <RoleItem
                role={role}
                key={role.id}
                onSelect={this.props.assignUser.bind(this, this.props.selectedUserId, role.id, this.props.selectedProjectId)}
              />
            )
        }

        {this.props.mode === 'default'
            ? <button className="assign" onClick={this.props.changeMode.bind(this, 'select-user')}>
              Assign more
            </button>
            : <button className="assign" onClick={this.props.changeMode.bind(this, null)}>
              Cancel
            </button>
        }
      </div>
    )
  }
}

const mapStateToProps = state => ({
  filteredUsers: state.filteredUsers,
  isFetching: state.isFetching,
  mode: state.mode,
  projects: state.projects,
  roles: state.roles,
  selectedProjectId: state.selectedProjectId,
  selectedUserId: state.selectedUserId,
  users: state.users,
})

const mapDispatchToProps = dispatch => ({
  assignUser: (userId, roleId, projectId) => {dispatch(actions.assignUser(userId, roleId, projectId))},
  changeMode: mode => {dispatch(actions.changeMode(mode))},
  changeProject: id => {dispatch(actions.changeProject(id))},
  fetchData: () => {dispatch(actions.fetchData())},
  selectUser: userId => {dispatch(actions.selectUser(userId))},
  unassignUser: (userId, projectId) => {dispatch(actions.unassignUser(userId, projectId))},
})

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(RolesWidget)

