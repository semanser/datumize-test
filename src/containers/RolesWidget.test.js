import React from 'react'
import ReactDOM from 'react-dom'
import ShallowRenderer from 'react-test-renderer/shallow'
import { RolesWidget } from './RolesWidget'

const renderer = new ShallowRenderer()

const props = {
  projects: [
    {
      id: 1,
      name: 'Test project 1',
    },
    {
      id: 2,
      name: 'Test project 2',
    },
  ],
  users: [
    {
      id: 1,
      name: 'test user 1',
    },
    {
      id: 2,
      name: 'test user 2',
    },
  ],
  selectedProjectId: 1,
  selectedUserId: 1,
  assignUser: () => {},
  changeMode: () => {},
  selectUser: () => {},
  unassignUser: () => {},
  filteredUsers: [
    {
      id: 1,
      name: 'test user 1',
    },
  ],
  roles: [
    {
      id: 1,
      name: 'Admin',
    },
    {
      id: 2,
      name: 'Editor',
    }
  ],
}

describe('RolesWidget', () => {
  it('Renders component without crashing', () => {
    const div = document.createElement('div')
    const tree = renderer.render(<RolesWidget {...props}/>)
    expect(tree).toMatchSnapshot()
  })

  it('should render loader if props.isFetching', () => {
    const div = document.createElement('div')
    const tree = renderer.render(<RolesWidget {...Object.assign({}, props, { isFetching: true })}/>)
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with props.mode === "default"', () => {
    const div = document.createElement('div')
    const tree = renderer.render(<RolesWidget {...Object.assign({}, props, { mode: "default" })}/>)
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with props.mode === "select-user"', () => {
    const div = document.createElement('div')
    const tree = renderer.render(<RolesWidget {...Object.assign({}, props, { mode: "select-user" })}/>)
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with props.mode === "select-role"', () => {
    const div = document.createElement('div')
    const tree = renderer.render(<RolesWidget {...Object.assign({}, props, { mode: "select-role" })}/>)
    expect(tree).toMatchSnapshot()
  })
})

