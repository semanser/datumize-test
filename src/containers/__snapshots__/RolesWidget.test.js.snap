// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RolesWidget Renders component without crashing 1`] = `
<div
  className="widget"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      Test project 1
    </div>
    <div
      className="description"
    >
      Now select the role for test user 1
    </div>
  </div>
  <button
    className="assign"
    onClick={[Function]}
  >
    Cancel
  </button>
</div>
`;

exports[`RolesWidget should render correctly with props.mode === "default" 1`] = `
<div
  className="widget"
>
  <RolesSelect
    assignUser={[Function]}
    changeMode={[Function]}
    filteredUsers={
      Array [
        Object {
          "id": 1,
          "name": "test user 1",
        },
      ]
    }
    mode="default"
    projects={
      Array [
        Object {
          "id": 1,
          "name": "Test project 1",
        },
        Object {
          "id": 2,
          "name": "Test project 2",
        },
      ]
    }
    roles={
      Array [
        Object {
          "id": 1,
          "name": "Admin",
        },
        Object {
          "id": 2,
          "name": "Editor",
        },
      ]
    }
    selectUser={[Function]}
    selectedProjectId={1}
    selectedUserId={1}
    unassignUser={[Function]}
    users={
      Array [
        Object {
          "id": 1,
          "name": "test user 1",
        },
        Object {
          "id": 2,
          "name": "test user 2",
        },
      ]
    }
  />
  <UserItem
    onUnassign={[Function]}
    user={
      Object {
        "id": 1,
        "name": "test user 1",
      }
    }
  />
  <button
    className="assign"
    onClick={[Function]}
  >
    Assign more
  </button>
</div>
`;

exports[`RolesWidget should render correctly with props.mode === "select-role" 1`] = `
<div
  className="widget"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      Test project 1
    </div>
    <div
      className="description"
    >
      Now select the role for test user 1
    </div>
  </div>
  <RoleItem
    onSelect={[Function]}
    role={
      Object {
        "id": 1,
        "name": "Admin",
      }
    }
  />
  <RoleItem
    onSelect={[Function]}
    role={
      Object {
        "id": 2,
        "name": "Editor",
      }
    }
  />
  <button
    className="assign"
    onClick={[Function]}
  >
    Cancel
  </button>
</div>
`;

exports[`RolesWidget should render correctly with props.mode === "select-user" 1`] = `
<div
  className="widget"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      Test project 1
    </div>
    <div
      className="description"
    >
      Select the user below
    </div>
  </div>
  <UserItem
    onSelect={[Function]}
    user={
      Object {
        "id": 1,
        "name": "test user 1",
      }
    }
  />
  <button
    className="assign"
    onClick={[Function]}
  >
    Cancel
  </button>
</div>
`;

exports[`RolesWidget should render loader if props.isFetching 1`] = `
<div
  className="widget"
>
  <GridLoader
    className="loader"
    color="#5E85E7"
    loading={true}
    margin="2px"
    size="10px"
  />
</div>
`;
